[6.7] EZP-29139: RegenerateUrlAliasesCommand should keep historized Url Aliases

alongosz opened the pull request:
| Question           | Answer
| ------------------ | ------------------
| **JIRA issue**     | [EZP-29138](https://jira.ez.no/browse/EZP-29138)
| **Related PR** | #2429
| **Bug/Improvement**| yes
| **New feature**    | yes
| **Target version** | `5.4`, `6.x`+
| **BC breaks**      | yes
| **Tests pass**     | yes
| **Doc needed**     | yes

## Summary
The `ezplatform:urls:regenerate-aliases` command is a replacement for `ezplatform:regenerate:legacy_storage_url_aliases`.

The main difference is that the new command entirely relies on new APIs to regenerate URL aliases and thus it supports archiving URL aliases and redirecting archived ones to the new ones (a.k.a historization).

## Backporting
This is a `6.7` (aimed for `5.4`) backport, because we consider lack of the proper archiving of URL aliases as a bug. However to keep BC old command is kept as the current one uses entirely new approach.

## BC
PR introduces several missing API features to be able to implement archiving of URL aliases. While this might be considered as a BC break, it will not affect API Consumer and it is necessary due to architecture issues.

## Approach
This approach utilizes bulk loading of Location and Content Value Objects to work around known memory leaks on the lowest (`PDO`) layer. The approach presented in #2388 proved to be too consuming on resources for large databases (both memory and time-wise). While in this approach we cannot completely avoid memory leaks, the improvement is significant.

## Reviewers
Some parts of the code, related to loading in bulk Content and Content Info List were backported from newer versions, so the code is "as-is" to minimize merge-up conflicts.

**TODO**:
- [x] Remove corrupted URL alias entry from `cleandata.sql`
- [x] Backport SPI `Content\Handler::loadContentInfoList`.
- [x] Introduce API to load all Locations in a paginated way.
- [x] Backport SPI to load multiple Content items
- [x] Introduce API to load multiple Content items
- [x] Backport SPI archiving of deleted Translations.
- [x] Introduce SPI to delete corrupted entries
- [x] Implement command utilizing bulk loading of API Value Objects.
- [x] Deprecate the `ezplatform:regenerate:legacy_storage_url_aliases` command.
- [x] Fix new code according to Coding Standards (`$ composer fix-cs`).
- [x] Ask for Code Review.

After QA:
- [x] Move logic deleting corrupted entries before regenerating system URL aliases (2a7a88a).
- [x] Improve error message coming from URLAlias Gateway `loadPathData` method to hint to use command introduced by this PR to fix the problem (a3eccd2).
- [x] Change `RuntimeException` to API Badstate exception in the mentioned `loadPathData` method (a3eccd2).
- [x] `refreshSystemUrlAliasesForLocation` API: catch every exception inside refreshing to make sure rollback is always executed (arbitrary exception could come from a custom Field Type implementation) (32ee937).

**TODO on merge to 7.2/master**:
- [ ] Change loading of content to `$location->getContent()` which will handle bulk loading for us and simplify code
- [ ] Adapt SiteAccessAware repo for new API's

Followup PR _(6.7 or master depending)_ to improve the backported code:
- [ ] From Adam: [Improve the query selection logic](https://github.com/ezsystems/ezpublish-kernel/pull/2423#discussion_r213831234)

adamwojs commented on eZ/Publish/Core/Persistence/Legacy/Content/UrlAlias/Gateway/DoctrineDatabase.php:
```diff
@@ -1132,4 +1158,225 @@ public function getLocationContentMainLanguageId($locationId)
+        );
+
+        $connection = $this->dbHandler->getConnection();
+        /** @var \Doctrine\DBAL\Connection $connection */
```
Nitpick: This comment should be on top of previous line

adamwojs commented on eZ/Publish/Core/Persistence/Legacy/Content/UrlAlias/Gateway/DoctrineDatabase.php:
```diff
@@ -1132,4 +1158,225 @@ public function getLocationContentMainLanguageId($locationId)
+    /**
+     * Get DBMS-specific integer type.
+     *
+     * @param $databasePlatform
```
NItpick: Missing type hint for `$databasePlatform` parameter

adamwojs commented on eZ/Publish/Core/Persistence/Legacy/Content/UrlAlias/Handler.php:
```diff
@@ -1004,4 +1013,66 @@ protected function getHash($text)
+    {
+        $this->transactionHandler->beginTransaction();
+        try {
+            $totalCount = 0;
```
Why `$totalCount` is intialized with 0 instead of `$this->gateway->deleteUrlAliasesWithoutLocation()` ?

alongosz replied:
For readability.

alongosz replied:
Fixed anyway ;)

adamwojs commented on eZ/Publish/Core/Persistence/Legacy/Content/UrlAlias/Handler.php:
```diff
@@ -1004,4 +1013,66 @@ protected function getHash($text)
+                    }
+                )
+            );
+        }
```
Logic behind lines 1027-1037 could be written as well using nested loops: 

```php
foreach ($urlAliases as $urlAlias) {
    foreach ($urlAlias->languageCodes as $languageCode) {
        if (!in_array($languageCode, $languageCodes)) {
            $removedLanguages[] = $languageCode;
        }
    }
}
```

Am I right?

alongosz replied:
Yeah, that's more readable. Thanks.

alongosz commented on eZ/Publish/Core/Persistence/Legacy/Content/Gateway/DoctrineDatabase.php:
```diff
@@ -850,36 +851,119 @@ public function load($contentId, $version, array $translations = null)
+     *
+     * @return array
+     */
+    private function internalLoadContentInfo(DoctrineQueryBuilder $query)
```
@adamwojs this can be handled on master only as this is part of the backport from there. Otherwise we'll end up with difficult conflict on merge up

andrerom replied:
Lets keep track of things we want to improve in master as a list in PR description here on section on  "As-Is" note so we all know and build up things we want to cleanup in second pass. works?

andrerom replied:
Lets keep track of things we want to improve in master as a list in PR description here on section on  "As-Is" note so we all know and build up things we want to cleanup in second pass. Works?

alongosz commented on eZ/Publish/Core/Persistence/Legacy/Content/Gateway/DoctrineDatabase.php:
```diff
@@ -926,21 +1014,16 @@ public function loadContentInfo($contentId)
-        $row = $this->internalLoadContentInfo($query);
+        $query = $this->connection->createQueryBuilder();
+        $query->where('c.remote_id = :id')
+              ->setParameter('id', $remoteId, \PDO::PARAM_STR);
```
Hmm, this is a backport, but might be good idea to fix it on merge up anyway. Done.

alongosz commented on eZ/Bundle/EzPublishCoreBundle/Command/RegenerateUrlAliasesCommand.php:
```diff
@@ -0,0 +1,232 @@
+                '<info>Found %d Locations. Regenerating System URL aliases...</info>',
+                $locationsCount
+            )
+        );
```
Good idea. Changed.

alongosz commented:
@adamwojs fixed what I could :)

andrerom commented on eZ/Bundle/EzPublishCoreBundle/Command/RegenerateUrlAliasesCommand.php:
```diff
@@ -0,0 +1,241 @@
+            )
+            ->setHelp(
+                <<<EOT
+<error>Make sure to back up your database before proceeding.</error>
```
nitpick: warning? But this can work as well, just a but strange to see usage of error on something which is not a error condition :P

andrerom replied:
Maybe also group the two errors/warnings togheter at the top.

For instance:
```
Before you continue:
- Make sure to back up your database before proceeding.
- Take installation offline, during the script execution the database should not be modified.
```

alongosz replied:
> nitpick: warning? But this can work as well, just a but strange to see usage of error on something which is not a error condition :P

Its about the color used for color-capable terminals. There's no `warning` formatting tag. See [Using color styles](https://symfony.com/doc/2.8/console/coloring.html#using-color-styles) Sf doc. We could use some other tag, but red seems like the best choice chere.

> Maybe also group the misc errors/warnings together at the top.

Good idea. I expanded on it and reused the same information during command runtime, before user decides to continue processing Locations. Fixed via 01bb253.

andrerom commented on eZ/Bundle/EzPublishCoreBundle/Command/RegenerateUrlAliasesCommand.php:
```diff
@@ -0,0 +1,241 @@
+            );
+            $this->processLocations($locations, $progressBar);
+            // force garbage collection
+            gc_collect_cycles();
```
Afaik it's fixed in PHP 7.3 but could be you get some extra perf by disabling gc within each iteration, ref:
https://blog.blackfire.io/performance-impact-of-the-php-garbage-collector.html
https://tideways.com/profiler/blog/how-to-optimize-the-php-garbage-collector-usage-to-improve-memory-and-performance

alongosz replied:
> Afaik it's fixed in PHP 7.3 but could be you get some extra perf by disabling gc within each iteration, ref:
> https://blog.blackfire.io/performance-impact-of-the-php-garbage-collector.html
> https://tideways.com/profiler/blog/how-to-optimize-the-php-garbage-collector-usage-to-improve-memory-and-performance

Hmm, so I investigated it further today and I don't see any increase in performance both time and memory-wise. But maybe with PHP 7.3 it will make a difference. Changed via 4a584bc.

andrerom commented on eZ/Bundle/EzPublishCoreBundle/Command/RegenerateUrlAliasesCommand.php:
```diff
@@ -0,0 +1,241 @@
+                $locations
+            ),
+            function (ContentInfo $contentInfo) {
+                return $contentInfo->isPublished();
```
Is this to prepare for location drafts feature? Afaik only published content will be returned when we query via locations. Or did you get issues before you added this and filter below for published flag?

alongosz replied:
No, I had in mind drafts with assigned location but not published yet. However indeed `ezcontentobject_tree` should have only published ones, my mistake. Removed this check also via 01bb253.

andrerom commented on eZ/Publish/API/Repository/URLAliasService.php:
```diff
@@ -130,4 +131,19 @@ public function reverseLookup(Location $location, $languageCode = null);
+     * @param \eZ\Publish\API\Repository\Values\Content\Location $location
+     * @param \eZ\Publish\API\Repository\Values\Content\Content $content pre-loaded Content for the $location
+     */
+    public function refreshSystemUrlAliasesForLocation(Location $location, Content $content = null);
```
given we introduce `$location->getContent()` in 2.2, maybe we should:
A. Not document this (so internal argument..) and get rid of it in 7.2 merge
B. Directly document this argument as deprecated and remove in 7.2/master merge

alongosz replied:
> given we introduce $location->getContent() in 2.2, maybe we should:
> A. Not document this (so internal argument..) and get rid of it in 7.2 merge
> B. Directly document this argument as deprecated and remove in 7.2/master merge

B. I added to PhpDoc info that it is deprecated (54ab4d0), because I don't like PhpStorm complaining about missing/erroneous doc.

andrerom commented on eZ/Publish/Core/Persistence/Legacy/Content/UrlAlias/Handler.php:
```diff
@@ -1004,4 +1013,61 @@ protected function getHash($text)
+        // map languageCodes to their IDs
+        $languageIds = array_map(
+            function ($languageCode) {
+                return $this->languageHandler->loadByLanguageCode($languageCode)->id;
```
todo, we need to have a multi load spi for this to get rid of these calls, but this can be marked for followup, I might be able to handle it end of september

andrerom replied:
followup is fine

alongosz commented:
Note: while testing I found a flaw in my bulk loading with always available logic, fixed via d3a7425. If we want to use globally always available, after query optimizations, we will have to include every main language, no matter if specific Content item has it set or no, otherwise we end up with list of main languages for always available Content items only, which is not what we want. As a side effect we might not get all the languages, but I think it's better that way.

So to get all Content items in all languages, always available must be set to false (if we got all languages, the main one will be included anyway). Also fixed API usage via 01bb253.

alongosz commented:
Note: aligned with recent #2429 changes via 8dbd79a.

andrerom commented on eZ/Publish/API/Repository/ContentService.php:
```diff
@@ -148,6 +148,20 @@ public function loadContent($contentId, array $languages = null, $versionNo = nu
+     * @param \eZ\Publish\API\Repository\Values\Content\ContentInfo[] $contentInfoList
+     * @param string[] $languages A language priority, filters returned fields and is used as prioritized language code on
+     *                            returned value object. If not given all languages are returned.
+     * @param bool $useAlwaysAvailable Add Main language to \$languages if true (default) and if alwaysAvailable is true
```
Unless all languages have been asked for

alongosz replied:
Fixed doc via fixup 3093853

andrerom commented on eZ/Publish/Core/Repository/ContentService.php:
```diff
@@ -455,6 +455,55 @@ public function loadContentByRemoteId($remoteId, array $languages = null, $versi
+        $contentIds = [];
+        foreach ($contentInfoList as $contentInfo) {
+            $contentIds[] = $contentInfo->id;
+            if (!$useAlwaysAvailable) {
```
can also skip if language is all, see note on constant regarding that.

alongosz replied:
I replaced this block with [this logic](https://github.com/ezsystems/ezpublish-kernel/commit/10c2e8ff2bdee08599e5998eea0a2a215700707f#diff-4043c8073a9a8c2593b809f674fafa42R194) you've mentioned in other comment, so this would be redundant. Also changed via fixup 3093853

andrerom commented on eZ/Bundle/EzPublishCoreBundle/Command/RegenerateUrlAliasesCommand.php:
```diff
@@ -0,0 +1,238 @@
+                // load Content list in all languages
+                return $repository->getContentService()->loadContentListByContentInfo(
+                    $contentInfoList,
+                    [],
```
You can safely backport `Language::ALL` _(constant only)_ so this as well as other people's code can start to get forward compatible with 3.x _(or whenever we switch siteaccess awareness on by default)_.

ref: https://github.com/ezsystems/ezpublish-kernel/blob/master/eZ/Publish/API/Repository/Values/Content/Language.php#L23-L26

alongosz replied:
Added via fixup: de332cf

andrerom commented on eZ/Publish/Core/Repository/ContentService.php:
```diff
@@ -455,6 +455,55 @@ public function loadContentByRemoteId($remoteId, array $languages = null, $versi
+                continue;
+            }
+
+            if (!in_array($contentInfo->mainLanguageCode, $languages)) {
```
The previous logic was more correct afaik, at lest that is what I ended up with in 7.2: 
https://github.com/ezsystems/ezpublish-kernel/commit/10c2e8ff2bdee08599e5998eea0a2a215700707f#diff-4043c8073a9a8c2593b809f674fafa42R194


Also remember to use `array_unique` on the resulting array :)

andrerom commented:
From my side this is all fine now, so might be ready for approval or last round of reviews by @adamwojs  others ;)

alongosz commented:
Changes after QA:
- Moved logic deleting corrupted entries before regenerating system URL aliases (2a7a88a).
- Improved error message coming from URLAlias Gateway `loadPathData` method to hint to use command introduced by this PR to fix the problem. Also changed `RuntimeException` to API `BadStateException` (a3eccd2).
- `refreshSystemUrlAliasesForLocation` API: catch every exception inside refreshing to make sure rollback is always executed (arbitrary exception could come from a custom Field Type implementation) (32ee937).

alongosz commented:
Note: Merging this so it goes into `7.3.0-beta1` after merge up. It means that the second round of QA will have to happen after merge. // cc @mnocon
